### DISCUSSION TITLE
GraphQL Support

### DISCUSSION THREAD

**Discussion by nitech** [craftcms/commerce#2350] [Ideas]
I need to be able to CRUD products in Craft Commerce via an API. Either REST or GraphQL. Are there any plans for implementing this? 

I see Craft CMS has an issue https://github.com/craftcms/cms/issues/4835 at the 4.0 milestone. I hope that also includes mutations on Craft Commerce.

**Comment by lukeholder** [comment]
Yes, plans to have support for GraphQL in Commerce in 3.0 this year - mutation support coming to the GraphQL API after that.

**Reply by nitech** [reply]
What's the status of mutations for Products, Users and Orders as of today @lukeholder?

**Comment by lukeholder** [comment]
- [x] Products
- [x] Variants
- [ ] Orders

**Reply by devonmather-mymy** [reply]
Is there an ETA of being able to fetch orders?

**Reply by Ferre-Uhawha** [reply]
Is it possible to make a reactive cart using GraphQL? Or is Alpine.js required?

**Comment by nitech** [comment]
Great work guys. So if I understand this correctly, Craft Commerce 3.0 has GraphQL read-only support for all Commerce-entities? 

If then, do you have any plans for when mutations will come? 

Keep up the good work!

**Comment by lukeholder** [comment]
@nitech Read only support for products. Will extend that to variants and orders when we can.

**Comment by timkelty** [comment]
@lukeholder any update on variants/orders?

**Comment by lukeholder** [comment]
@timkelty Variant support coming in 3.1, orders after that.

**Comment by dnplkndll** [comment]
we need to know how to load the products and update the order status. Was hoping that the mutation support would be possible to add the product/variants and update the orders from the erp. 
Is this in the works?

**Comment by lukeholder** [comment]
Mutations just landed in craft core in 3.5 a couple weeks ago. We need to first add mutation support for Products and Variants.

Then when Craft's core GraphQL supports gets per-session/user permissions we will look to add Order support.

**Comment by terraindesign** [comment]
+1 for GraphQL support for orders. Would love to try and integrate this with clients accounting system to pull received orders, update order status. Our current system is a bit clumsy.

**Comment by tomduncalf** [comment]
Is there an expected timeline on Products mutations? I'd love to be able to use Craft for an upcoming project but need some way to automate managing products. There's no REST API is there?

**Comment by controlnocontrol** [comment]
That's a really nice feature!

2 questions:

- how to we query Commerce products?  Any examples/documentation online?
- when can we expect CUD for products at least?

**Comment by karensg** [comment]
What would be the best way to get orders out of Craft. As I understand the orders are not part of GraphQL yet.
Is using element-api the way to go?

**Reply by nfourtythree** [reply]
Hi @karensg 

You are correct, at the moment orders are not part of GraphQL. You could definitely use the [Element API](https://plugins.craftcms.com/element-api) plugin, you would just need to make sure you account for the possible security concerns e.g. someone being able to access another customer's information.

Thanks!

**Reply by nicholashu** [reply]
@karensg I've come across this issue before. I created a craft module that returned the data i wanted.

a simple function like  this will give you the orders for a specific customer

```
public function actionGetOrders()
    {
        $user = Craft::$app->getUser();
      
        // return error if not logged in
        $this->handleError($user->getIsGuest(), 'NOT_LOGGED_IN');

        // find matching order
        $orders = Order::find()->user($user->id)->isCompleted()->all();

         // return error if no orders
        $this->handleError(!$orders, 'NO_ORDER');

        return $orders;
    }
```
you could then return the order data via a post request to the action or create a custom graphql query created in the same module. Setting up a custom graphql endpoint is a little annoying as you need to create the generator, interface, service and types. But, if the rest of the frontend is using GraphQL, I would prefer to use GraphQL for retrieving the data.

**Reply by vitalijalbu** [reply]
why custom php code?? this mut be integrated as default, since you pay 999$ for the plugin. am i wrong??

**Comment by andreasottosson** [comment]
Any updates on this? Will you also enable mutations to create a checkout etc?

**Comment by smockensturm** [comment]
Anything at all to report here @lukeholder ?

**Reply by lukeholder** [reply]
I've responded below.

**Comment by lukeholder** [comment]
@smockensturm @andreasottosson It's on the list but we don't have a timeline on GraphQL support for checkout. We could look to get read/query support for orders sooner, but again we can't commit to a timeline sorry.

**Reply by roberttolton** [reply]
@lukeholder Has there been any movement on this? Currently a show-stopper (need to get / update Orders data) for a big new project. Thanks!

**Reply by jesuismaxime** [reply]
+1 We have a potential project planned next year and since Craft is our go to but we moved 100% to headless mode; having the Commerce experience into the same App would be great if not required by our client!

**Reply by timkelty** [reply]
While maybe not ideal, it is worth noting traditional REST/JSON API endpoints exist for cart operations, so it is still possible to go fully headless.

**Reply by vitalijalbu** [reply]
passed 1 year and still don't have any response... great job... 👏

**Comment by vitalijalbu** [comment]
Hi all, how come commerce costs 999$ but you even don't have default crud for products, variants, orders, checkout and so on... Nowdays all is headless, so in my opinion all these factors are a must ! @lukeholder am I wrong?

**Comment by lukeholder** [comment]
@vitalijalbu 

We are sorry that mutations for Orders over the GraphQL API has not yet arrived. The cart and checkout controllers all support JSON requests/responses to modify the cart, and can be used from a headless setup, although we know that is not ideal.

I have placed it officially on the public roadmap for Commerce 5, which will be out in beta later this year.

**Reply by vitalijalbu** [reply]
Take a look please into **saleor**: https://github.com/saleor/saleor
If they do it, I think also it's possible to do it in craft-commerce.

**Comment by timkelty** [comment]
@vitalijalbu To clarify, the core of this issue is not really a Commerce issue at heart. The current state of Craft's GQL API uses static API keys to grant schema access. For something like cart mutations to make sense, we need the ability to authenticate via GQL instead of a public token.

Craft 5 has some big improvements in the works around authentication, which should get this going in the right direction, so stay tuned.

Discussion: https://github.com/craftcms/commerce/discussions/2350#discussioncomment-6109850

**Reply by vitalijalbu** [reply]
So basically if I'd need asap to start a commerce project, I have 2 options:
- wait for v5
- or choose some craft alternatives (maybe something that already have all functionalities) like saleor

Personally I'd like craft and don't want to go with any other stacks, but I'm forced to... and also don't want to code by myself all endpoints for a full checkout proccess...

**Reply by lukeholder** [reply]
> and also don't want to code by myself all endpoints for a full checkout process

Just confirming you understand that all commerce controller/action endpoints are headlessly (json enabled) capable, so this shouldn't be necessary.

**Reply by devonmather-mymy** [reply]
Is this still on the roadmap for Commerce 5? I see the docs for V5 seem to be out, but no mention of GraphQL 👀 https://craftcms.com/docs/commerce/5.x/

**Comment by nitech** [comment]
@lukeholder any news on this? I don't understand why the default authentication mechanism is insufficient. What I and a lot of the posters here want is the ability to communicate between the ERP or CRM-system and Craft, so that when an order gets a tracking number or a status update from the ERP system (or elsewhere), we can update the order in Craft - or when we do changes to available items in storage for a product in the Warehouse Management System, we can update the availability in Craft Commerce. 

So I am not after a headless Craft Commerce - I just want to communicate between my systems.

**Reply by devonmather-mymy** [reply]
➕ 1 for this requirement

**Comment by vitalijalbu** [comment]
news guys about commerce cart mutations??

**Reply by lukeholder** [reply]
We are working on it but we can't give a timeline sorry!

**Reply by denisyilmaz** [reply]
@lukeholder any updates on this?

**Reply by lukeholder** [reply]
Will be a Craft 6 and Commerce 6 thing. GQL does not have user level permissions for mutations in core at the moment.
